src/mappings/entities/mapSettlement.ts

```diff
@@ -0,0 +1,203 @@
+import { Codec } from "@polkadot/types/types";
+import { SubstrateEvent } from "@subql/types";
+import { Settlement, Instruction } from "../../types";
+import { getSigner, getTextValue, hexToString, serializeTicker } from "../util";
+import { EventIdEnum, ModuleIdEnum } from "./common";
+
+enum SettlementResultEnum {
+  None = "None",
+  Executed = "Executed",
+  Failed = "Failed",
+  Rejected = "Rejected",
+}
+
+enum InstructionStatusEnum {
+  Created = "Created",
+  Executed = "Executed",
+  Rejected = "Rejected",
+  Failed = "Failed",
+}
+
+const updateEvents: EventIdEnum[] = [
+  EventIdEnum.InstructionAuthorized,
+  EventIdEnum.InstructionUnauthorized,
+  EventIdEnum.InstructionAffirmed,
+  EventIdEnum.InstructionFailed,
+];
+
+const finalizedEvents: EventIdEnum[] = [
+  EventIdEnum.InstructionExecuted,
+  EventIdEnum.InstructionRejected,
+  EventIdEnum.InstructionFailed,
+];
+
+// Translates events into a settlements table. This inlcudes transfers between a users portfolio combined with completed Instructions.
+export async function mapSettlement(
+  blockId: number,
+  eventId: EventIdEnum,
+  moduleId: string,
+  params: Codec[],
+  event: SubstrateEvent
+): Promise<void> {
+  if (
+    moduleId === ModuleIdEnum.Portfolio &&
+    eventId === EventIdEnum.MovedBetweenPortfolios
+  ) {
+    await handlePortfolioMovement(blockId, eventId, params, event);
+  }
+
+  if (moduleId === ModuleIdEnum.Settlement) {
+    if (eventId === EventIdEnum.InstructionCreated) {
+      await handleInstructionCreated(blockId, eventId, params, event);
+    }
+
+    if (updateEvents.includes(eventId)) {
+      await handleInstructionUpdate(params, event);
+    }
+
+    if (finalizedEvents.includes(eventId)) {
+      await handleInstructionFinalizedEvent(blockId, eventId, params, event);
+    }
+  }
+}
+
+async function handlePortfolioMovement(
+  blockId: number,
+  eventId: EventIdEnum,
+  params: Codec[],
+  event: SubstrateEvent
+) {
+  const signer = getSigner(event.extrinsic);
+  const identityId = getTextValue(params[0]);
+  const from = JSON.parse(params[1].toString());
+  const to = JSON.parse(params[2].toString());
+  const ticker = serializeTicker(params[3]);
+  const amount = getTextValue(params[4]);
+
+  const legs = [
+    {
+      ticker,
+      amount,
+      from: {
+        did: from.did,
+        number: from.kind.user ? from.kind.user : 0,
```

Comment:
nit pick -> from.kind.user || 0
You can ignore this as well. I prefer the || whenever kind of using it for setting default value